### PR TITLE
Circuitry upgrade Pt 2

### PR DESCRIPTION
**Pull request by arsserpentarium**
1.All reagent containers are acidproof now and have pin with self reference(to make it more easy to link them)
2.reagent scaner. It's like a beaker. But can give a list of contained reagents.
3.reagent filter. Pump,that can transfer only desired reagents,or all,except unwanted.
4.fuel cell.Converts chemicals to energy.
5.2 pins memory circuit. Sometimes it is very useful.
6.pulse demultiplexers. A lot like regular demultiplexer. But it doesn't distributes data. Just sends pulse to given activator pin.
7.list operations. choose element by index. write element by index. delete element by index search desired element in list. lenght of list.
8.strings: find element in string. explode string by delimeter.

### CODE REVIEW

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 1]
```diff
@@ -105,9 +107,11 @@
 	push_data()
 	activate_pin(2)
 
+
+
 /obj/item/integrated_circuit/input/adv_med_scanner
 	name = "integrated advanced medical analyser"
-	desc = "A very small version of the common medical analyser.  This allows the machine to know how healthy someone is.  \
+	desc = "A very small version of the medbot's medical analyser.  This allows the machine to know how healthy someone is.  \
```
Comment (Belsima) [inline code/modules/integrated_electronics/subtypes/input.dm]: medibot*

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 2]
```diff
@@ -130,7 +134,7 @@
 	var/mob/living/carbon/human/H = get_pin_data_as_type(IC_INPUT, 1, /mob/living/carbon/human)
 	if(!istype(H)) //Invalid input
 		return
-	if(H.Adjacent(get_turf(src))) // Like normal analysers, it can't be used at range.
+	if(H in view(get_turf(H))) // Like medbot's analyzer it can be used in range..
```
Comment (Belsima) [inline code/modules/integrated_electronics/subtypes/input.dm]: medibot*

File: code/modules/integrated_electronics/passive/power.dm [hunk 3]
```diff
@@ -29,6 +29,28 @@
 		if(assembly)
 			assembly.give_power(adjusted_power)
 
+/obj/item/integrated_circuit/passive/power/starter
+	name = "starter"
+	desc = "This tiny circuit will send pulse right after device is turned on. Or when power is restored."
```
Comment (Belsima) [inline code/modules/integrated_electronics/passive/power.dm]: a pulse

File: code/modules/integrated_electronics/passive/power.dm [hunk 4]
```diff
@@ -82,6 +104,44 @@
 	origin_tech = list(TECH_POWER = 3, TECH_ENGINEERING = 3, TECH_DATA = 2)
 	spawn_flags = IC_SPAWN_RESEARCH
 	var/power_amount = 250
+//fuel cell
+
+/obj/item/integrated_circuit/passive/power/chemical_cell
+	name = "fuel cell"
+	desc = "Produces electricity from chemicals."
+	icon_state = "chemical_cell"
+	extended_desc = "This is effectively an internal beaker.It will consume and produce power from phoron, slime jelly, welding fuel, carbon,\
```
Comment (Belsima) [inline code/modules/integrated_electronics/passive/power.dm]: Need space after 'beaker', and no space after 'blood'.

**Comment by arsserpentarium** [comment]
fixed

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 5]
```diff
@@ -134,7 +136,9 @@
 	var/mob/living/carbon/human/H = get_pin_data_as_type(IC_INPUT, 1, /mob/living/carbon/human)
 	if(!istype(H)) //Invalid input
 		return
+
 	if(H in view(get_turf(H))) // Like medbot's analyzer it can be used in range..
```
Comment (Belsima) [inline code/modules/integrated_electronics/subtypes/input.dm]: medibot too, but that might be for a seperate PR

Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/input.dm]: Oh, it really neccesary to fix such things in comments?

Comment (Belsima) [reply code/modules/integrated_electronics/subtypes/input.dm]: Not really- just pointing it out.

Comment (Belsima) [reply code/modules/integrated_electronics/subtypes/input.dm]: It isn't- I just usually roll it in with a commit if another needs to be made for other things.

File: code/modules/integrated_electronics/subtypes/lists.dm [hunk 6]
```diff
@@ -52,12 +52,123 @@
...
+	var/index = get_pin_data(IC_INPUT, 2)
+	var/item = get_pin_data(IC_INPUT, 3)
+	input_list[index] = item
+	set_pin_data(IC_OUTPUT, 1, input_list)
+	push_data()
+	activate_pin(2)
+
+obj/item/integrated_circuit/list/len
+	name = "len circuit"
+	desc = "This circuit will give lenght of the list."
```
Comment (Belsima) [inline code/modules/integrated_electronics/subtypes/lists.dm]: Length.

File: code/modules/integrated_electronics/subtypes/memory.dm [hunk 7]
```diff
@@ -40,8 +40,15 @@
 		O.push_data()
 	activate_pin(2)
 
+/obj/item/integrated_circuit/memory/tiny
+	name = "small memory circuit"
+	desc = "This circuit can store two pieces of data."
+	icon_state = "memory4"
```
Comment (Belsima) [inline code/modules/integrated_electronics/subtypes/memory.dm]: This should really have a memory2 icon.

File: code/modules/integrated_electronics/subtypes/reagents.dm [hunk 8]
```diff
@@ -150,4 +269,95 @@
...
+		cont += RE.id
+	set_pin_data(IC_OUTPUT, 3, cont)
+	push_data()
+
+
+/obj/item/integrated_circuit/reagent/filter
+	name = "reagent filter"
+	desc = "Filtering liquids by list of desired or unwanted reagents."
+	icon_state = "reagent_filter"
+	extended_desc = "This is a filter, which will move liquids from the source ref to the target ref. \
```
Comment (Belsima) [inline code/modules/integrated_electronics/subtypes/reagents.dm]: Unnecessary comma.

**Comment by arsserpentarium** [comment]
Had some troubles with git.looks like it works now.Fixed all, what have been asked to fix.

**Comment by Anewbe** [comment]
I'm not sure why roughly half of the files that were touched in this got touched at all.

**Comment by Belsima** [comment]
I can't even tell what changed.

**Comment by arsserpentarium** [comment]
I've changed everything, what I've been asked to change.
Yes, there was uploaded some files wihout changes.  It was some glitch of github.

**Comment by Anewbe** [comment]
@arsserpentarium Yeah, I'm not comfortable merging this with all those strange file changes. If you want this to get in, I'm going to have to ask you to remake it.

**Comment by arsserpentarium** [comment]
I'll remake it tomorrow. But please, reviev it faster.

**Pull request closed without merging**